### PR TITLE
set max request headers timeout

### PR DESCRIPTION
**Pull request by ramaraochavali**
Currently we do not set this and should be set at gateways to prevent from clients creating idle streams without [headers](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#envoy-v3-api-field-extensions-filters-network-http-connection-manager-v3-httpconnectionmanager-request-headers-timeout)
If it is incorrect to reuse the same timeout - we can pick some other value.
- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

**Comment by ymesika** [comment]
I'm thinking whether this should actually be controlled/configured by the users in the gateway configuration rather than hardcoding some arbitrary value.
Which will also allow disabling the timer (as it is today) for users who actually need it (can't think of a scenario though).

**Comment by ramaraochavali** [comment]
> I'm thinking whether this should actually be controlled/configured by the users in the gateway configuration rather than hardcoding some arbitrary value.

This is not arbitrary value.  This is the default request timeout we set for all routes (including gateways) if it is not specified  in virtual services. It matches the current default behaviour.

**Comment by ramaraochavali** [comment]
> Doesn't sidecar have the same issue?

It typically does not. This mainly needed for untrusted downstreams.

> Wondeing can streamIdleTimeout be used to prevent idle stream leak? It has a default value 5m

If it is completely idle - stream idle timeout would detect. But if a rogue client sends partial headers once in a while (breaking stream idleness), the connection would be blocked forever. That is the reason this timeout has been added in Envoy

**Comment by howardjohn** [comment]
> It typically does not. This mainly needed for untrusted downstreams.

I don't think there is an assumption that we trust downstreams traffic even in sidecars. The only thing we trust is their identity, but we don't have an knowledge of whether those identities are trusted to only send well-formed traffic ("we" here is istio project, in a specific users specific deployment they may have made that decision).

**Comment by hzxuzhonghu** [comment]
> I do not know if it is really needed for sidecars where clients are "known" by their identities.

Istio is to resolve zero-trust, think about the case there is no auth because client has no sidecar.

**Comment by hzxuzhonghu** [comment]
+1 for sidecar

Looks good to me, normally requestTimeout > requestHeaderTimeout, we set  a loose timeout for header looks good

**Comment by ramaraochavali** [comment]
Added for sidecar inbound also

**Comment by ramaraochavali** [comment]
Thinking about this more - we disable all timeouts at sidecar inbound. Not sure if it is a good idea to add this. So put on hold

**Comment by ramaraochavali** [comment]
After a second thought, I think it is not safe to add it for sidecars as we disable all timeouts. Only added for gateways. 
@howardjohn @hzxuzhonghu PTAL and see what you think?

**Comment by hzxuzhonghu** [comment]
>After a second thought, I think it is not safe to add it for sidecars as we disable all timeouts.

From security view, i think the risk is same for both.

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 1]
```diff
@@ -673,8 +673,9 @@ func buildGatewayConnectionManager(proxyConfig *meshconfig.ProxyConfig, node *mo
 			Uri:     true,
 			Dns:     true,
 		},
-		ServerName:          EnvoyServerName,
-		HttpProtocolOptions: httpProtoOpts,
+		ServerName:            EnvoyServerName,
+		RequestHeadersTimeout: features.DefaultRequestTimeout,
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/gateway.go:677]: see why request headers timeout is preferred over request timeout https://github.com/envoyproxy/envoy/issues/11427#issue-630222924

**Comment by ramaraochavali** [comment]
Closing this as it is getting difficult to get consensus on this

**Pull request closed without merging**